calicoctl/commands/common/version_mismatch.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calicoctl/v3/calicoctl/commands/clientmgr"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/options"
+)
+
+var VERSION string
+
+func CheckVersionMismatch(configArg, allowMismatchArg interface{}) {
+	if allowMismatch, _ := allowMismatchArg.(bool); allowMismatch {
+		log.Infof("Skip version mismatch checking due to '--allow-version-mismatch' argument")
+
+		return
+	}
+
+	cf, _ := configArg.(string)
+
+	client, err := clientmgr.NewClient(cf)
+	if err != nil {
+		// If we can't connect to the cluster, skip the check. Either we're running a command that
+		// doesn't need API access, in which case the check doesn't need to be run, or we'll
+		// fail on the actual command.
+		log.Infof("Skip version mismatch checking due to not being able to connect to the cluster")
+
+		return
+	}
+
+	ctx := context.Background()
+
+	ci, err := client.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err != nil {
+		var notFound cerrors.ErrorResourceDoesNotExist
+		if errors.As(err, &notFound) {
+			// ClusterInformation does not exist, so skip version check.
+			log.Infof("Skip version mismatch checking due to ClusterInformation not being present")
+
+			return
+		}
+		fmt.Fprintf(os.Stderr, "Unable to get Cluster Information to verify version mismatch: %s\n Use --allow-version-mismatch to override.\n", err)
+		os.Exit(1)
+	}
+
+	clusterv := ci.Spec.CalicoVersion
+	if clusterv == "" {
+		// CalicoVersion field not specified in the cluster, so skip check.
+		log.Infof("Skip version mismatch checking due to CalicoVersion not being set")
+
+		return
+	}
+
+	clusterv = strings.Split(clusterv, "-")[0]
+
+	clientv := strings.Split(VERSION, "-")[0]
+
+	if clusterv != clientv {
+		fmt.Fprintf(os.Stderr, "Version mismatch.\nClient Version:   %s\nCluster Version:  %s\nUse --allow-version-mismatch to override.\n", VERSION, clusterv)
+		os.Exit(1)
```

Comment:
nit: I know that exiting here is less code, but I think it's a little less clear from a glance if you were to see the `CheckVersionMismatch` function called that you would know it would prevent the rest of the code from running. I think the more common way is to have it return an error and then have each command pass that error until it finally gets handled in `calicoctl.go` (where it'll exit). I think it makes it easier for anyone using the function to reuse it and leaves us options to handle the error in the future separately.

Reply:
Got it, yeah it makes sense... I'll make the changes, thanks!